internal/controllers/cluster/cluster_controller.go

```diff
@@ -137,6 +143,22 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		}
 		if err := patchCluster(ctx, patchHelper, cluster, patchOpts...); err != nil {
 			reterr = kerrors.NewAggregate([]error{reterr, err})
+		} else {
+			// Don't emit the Events if the state hasn't changed
+			if preReconcilePhase != cluster.Status.GetTypedPhase() {
+				// Failed clusters should get a Warning event
+				if cluster.Status.GetTypedPhase() == clusterv1.ClusterPhaseFailed {
+					r.recorder.Eventf(cluster, corev1.EventTypeWarning, string(cluster.Status.GetTypedPhase()), "Cluster is %s: %s", string(cluster.Status.GetTypedPhase()), *cluster.Status.FailureMessage)
```

Comment:
nit
```suggestion
					r.recorder.Eventf(cluster, corev1.EventTypeWarning, string(cluster.Status.GetTypedPhase()), "Cluster phase is %s: %s", string(cluster.Status.GetTypedPhase()), *cluster.Status.FailureMessage)
```
Same below

Reply:
I had thought about adding it to `reconcilePhase` originally but the concern I had with that was that it could emit the event even if the `patchCluster` failed for some reason. I figured it'd be better to emit the event _after_ the `Cluster` resource has been fully patched by the `Reconcile` function. Thoughts?

Reply:
Regarding `Cluster is %s` vs `Cluster phase is %s`: Because the values of the `cluster.Status.GetTypedPhase()` is a verb I thought it read better without the word `phase`. In the code comments its been referred to as phase, state & status so it appeared to have multiple meanings depending on the author at the time. I'm fine with changing it to include the word `phase` but I thought it'd make more sense to the end-user that the Cluster is the thing being described.

Reply:
TL;DR; the current approach not only is not consistent with the rest of the codebase, but it seems not re-entrant to failures. I also see the attempt of "serializing changes" kind of in conflict with how the system behaves.

Re-entrancy:

What if the system fails to emit events after patching the object? 
if I got it right the system will not emit the events at the next reconciliation, or in other words, the current approach is not re-entrant to failures.

"serializing changes" : 

Generalizing a little bit, you are trying to make Kubernetes/etcd to act like a transactional system, which it is not. 

In this specific case, it means that the events can be generated twice or generated in bulk vs being generated in an exact sequence, but this isn't an issue from my PoV, because it is consistent with how the system works (level-triggered and not edge-triggered, or in other words, the system is designed to reach the desired state, not really to be a state machine with a finite set of steps in between state A and B). 

The last consideration is that the idea of "serializing changes" is not sustainable in complex controllers, think e.g to controllers creating MS/Machines, templates + emitting events along the way. This is why we prefer re-entrancy vs focusing on a sequence of operations.

Reply:
> In this specific case, it means that we should accept that the events can be generated twice or generated in bulk vs being generated in an exact sequence

I wasn't concerned about the occasional duplicate events, I was more concerned about emitting an event prior to the resource change being saved. The creation of the event shouldn't occur until after the patch has been sent to the api server.

> if I got it right the system will not emit the events at the next reconciliation, or in other words, the current approach is not re-entrant to failures.

It can't emit the event at every reconcile loop or we'd be spamming the etcd database with events. It should only be creating events when there is a change in the field value. If for some reason the k8s apiserver fails to accept the request to create an event there really isn't a way for us to keep retrying that scenario as the next reconcile loop won't see a change to the field's value.

> the current approach not only is not consistent with the rest of the codebase

But I understand what you are saying about this not being consistent with the rest of the code base. Let me go back and re-read the way the reconcile logic works, from the sound of your comments I think I may need to refresh some of my understanding.

Reply:
@fabriziopandini let me know if this new approach is consistent with what you were thinking.